File: Number_of_Islands/Number_of_Islands.py
def numIslands(grid):
    n = len(grid)
    m = len(grid[0])
    visited = [[False for i in range(n)] for j in range(m)]

    def numIslandsUtils(i,j,grid,visited):
        if grid[i][j] == "0" or visited[i][j] == True:
            return

        visited[i][j] = True
        if i+1 < len(grid):
            numIslandsUtils(i+1,j,grid,visited)
        if i-1 < len(grid):
            numIslandsUtils(i-1,j,grid,visited)
        if j+1 < len(grid):
            numIslandsUtils(i,j+1,grid,visited)
        if j-1 < len(grid):
            numIslandsUtils(i,j-1,grid,visited)
        
    ans = 0
    for i in range(n):
        for j in range(m):
            if grid[i][j] == '1' and visited[i][j]==False:
                ans+=1
                numIslandsUtils(i,j,grid,visited)


    

    return ans




# grid = [
#   ["1","1","1","1","0"],
#   ["1","1","0","1","0"],
#   ["1","1","0","0","0"],
#   ["0","0","0","0","0"]
# ]
grid =[["1","1","1","1","1","0","1","1","1","1"],["1","0","1","0","1","1","1","1","1","1"],["0","1","1","1","0","1","1","1","1","1"],["1","1","0","1","1","0","0","0","0","1"],["1","0","1","0","1","0","0","1","0","1"],["1","0","0","1","1","1","0","1","0","0"],["0","0","1","0","0","1","1","1","1","0"],["1","0","1","1","1","0","0","1","1","1"],["1","1","1","1","1","1","1","1","0","1"],["1","0","1","1","1","1","1","1","1","0"]]
print(numIslands(grid))

File: Number_of_Islands/Number_of_Islands.go
package main

import "fmt"

func main() {
	// grid := [][]string{
	// 	{"1", "1", "1", "1", "0"},
	// 	{"1", "1", "0", "1", "0"},
	// 	{"1", "1", "0", "0", "0"},
	// 	{"0", "0", "0", "0", "0"}}

	grid := [][]string{{"1", "1", "1", "1", "1", "0", "1", "1", "1", "1"},
		{"1", "0", "1", "0", "1", "1", "1", "1", "1", "1"},
		{"0", "1", "1", "1", "0", "1", "1", "1", "1", "1"},
		{"1", "1", "0", "1", "1", "0", "0", "0", "0", "1"},
		{"1", "0", "1", "0", "1", "0", "0", "1", "0", "1"},
		{"1", "0", "0", "1", "1", "1", "0", "1", "0", "0"},
		{"0", "0", "1", "0", "0", "1", "1", "1", "1", "0"},
		{"1", "0", "1", "1", "1", "0", "0", "1", "1", "1"},
		{"1", "1", "1", "1", "1", "1", "1", "1", "0", "1"},
		{"1", "0", "1", "1", "1", "1", "1", "1", "1", "0"}}

	fmt.Println(numIslands(grid))
}

func numIslands(grid [][]string) int {
	count := 0
	n := len(grid)
	m := len(grid[0])
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == "1" {
				count += 1
				numIslandsUtils(i, j, n, m, grid)
			}
		}
	}
	return count

}

func numIslandsUtils(i, j, n, m int, grid [][]string) {
	grid[i][j] = "0"
	if isValid(i+1, j, n, m, grid) {
		numIslandsUtils(i+1, j, n, m, grid)
	}
	if isValid(i, j+1, n, m, grid) {
		numIslandsUtils(i, j+1, n, m, grid)
	}
	if isValid(i-1, j, n, m, grid) {
		numIslandsUtils(i-1, j, n, m, grid)
	}
	if isValid(i, j-1, n, m, grid) {
		numIslandsUtils(i, j-1, n, m, grid)
	}

}

func isValid(i, j, n, m int, grid [][]string) bool {
	if i >= 0 && i < n && j >= 0 && j < m && grid[i][j] == "1" {
		return true
	}
	return false
}
